src/components/login/MagicLink.tsx

```diff
@@ -0,0 +1,154 @@
+import { materialCells } from '@jsonforms/material-renderers';
+import { JsonForms } from '@jsonforms/react';
+import { Alert, Button, Typography } from '@mui/material';
+import { ApiError } from '@supabase/supabase-js';
+import { useClient } from 'hooks/supabase-swr';
+import { isEmpty } from 'lodash';
+import { useSnackbar, VariantType } from 'notistack';
+import React, { useState } from 'react';
+import { FormattedMessage, useIntl } from 'react-intl';
+import {
+    defaultOptions,
+    defaultRenderers,
+    hideValidation,
+    showValidation,
+} from 'services/jsonforms';
+
+const MagicLink = () => {
+    const { enqueueSnackbar } = useSnackbar();
+    const supabaseClient = useClient();
+    const intl = useIntl();
+
+    const [showErrors, setShowErrors] = useState(false);
+    const [loading, setLoading] = useState(false);
+
+    const [submitError, setSubmitError] = useState<ApiError | null>(null);
+    const [formData, setFormData] = useState<{ email: string | undefined }>({
+        email: undefined,
+    });
+    const [formErrors, setFormErrors] = useState<any[] | undefined>([]);
+
+    const schema = {
+        properties: {
+            email: {
+                description: intl.formatMessage({
+                    id: 'login.email.description',
+                }),
+                minLength: 5,
+                type: 'string',
+            },
+        },
+        required: ['email'],
+        type: 'object',
+    };
+
+    const uiSchema = {
+        elements: [
+            {
+                label: intl.formatMessage({
+                    id: 'login.email.label',
+                }),
+                scope: `#/properties/email`,
+                type: 'Control',
+            },
+        ],
+        type: 'VerticalLayout',
+    };
+
+    const displayNotification = (id: string, variant: VariantType) => {
+        enqueueSnackbar(
+            intl.formatMessage({
+                id,
+            }),
+            {
+                anchorOrigin: {
+                    vertical: 'top',
+                    horizontal: 'center',
+                },
+                variant,
+            }
+        );
+    };
+
+    const handlers = {
+        submit: async (event: React.FormEvent<HTMLFormElement>) => {
+            event.preventDefault();
+            if (!isEmpty(formErrors) || isEmpty(formData.email)) {
+                setShowErrors(true);
+                return;
+            }
```

Comment:
Starting to like the pattern when using async of checking for errors and then returning if there are some.